FWCore/ParameterSet/python/Types.py

```diff
@@ -2058,6 +2059,23 @@ def testAllowed(self):
             p2 = PSet(aValue=optional.allowed(int32,PSet))
             p2.aValue = PSet(i = int32(3))
             self.assertEqual(p2.aValue.i.value(),3)
+            p2 = PSet(aValue=optional.allowed(int32,PSet))
+            p2.aValue = dict(i = int32(3))
+            self.assertEqual(p2.aValue.i.value(),3)
+
+            p2 = PSet(aValue=optional.allowed(int32,PSetTemplate(b=optional.int32)))
+            self.assertEqual(p2.dumpPython(),'cms.PSet(\n    aValue = cms.optional.allowed(cms.int32,cms.PSetTemplate(\n    b = cms.optional.int32\n))\n)')
+            p2.aValue = 2
+            self.assertEqual(p2.aValue.value(),2)
+            p2 = PSet(aValue=optional.allowed(int32,PSetTemplate(b=optional.int32)))
+            p2.aValue = PSet(b = int32(3))
+            self.assertEqual(p2.aValue.b.value(),3)
+            p2 = PSet(aValue=optional.allowed(int32,PSetTemplate(b=optional.int32)))
+            p2.aValue = dict(b = 3)
+            self.assertEqual(p2.aValue.b.value(),3)
+            self.assertEqual(p2.dumpPython(), 'cms.PSet(\n    aValue = cms.PSet(\n        b = cms.int32(3)\n    )\n)')
+            p2 = PSet(aValue=optional.allowed(int32,PSetTemplate(b=optional.int32)))
+            self.assertRaises(TypeError, lambda: setattr(p2, "aValue", dict(i = 1)) )
```

Comment:
Could you add a tests for
```py
# Assign something not int32 nor PSet, e.g.
p2.aValue = double(3.14)

# Assign non-int32 inside the PSet, e.g.
p2.aValue = PSet(b = double(3.14))
```
?

Such assignments should raise an Exception, right?

Reply:
The `assertRaises()` could be also
```suggestion
            self.assertRaises(TypeError, setattr, p2, "aValue", dict(i = 1) )
```
(I don't have strong preferences though)

Reply:
```python
p2.aValue = PSet(b=double(3.14))
```
does not throw. I think that is expected for a `PSetTemplate` since the code explicitly states the type and we assume that if you do that you know what you intended (e.g. the C++ code might allow multiple types).
```python
p2.aValue = dict(b = 3.14)
```
does throw as the value doesn't match for an int type.

Reply:
> ```python
> ```python
> p2.aValue = PSet(b=double(3.14))
> ```
>   
> does not throw. I think that is expected for a PSetTemplate since the code explicitly states the type and we assume that if you do that you know what you intended (e.g. the C++ code might allow multiple types).

Ok, makes sense. Would it be worth of adding such test anyway to demonstrate the behavior in that case?